File: source/c/deque_pointers.c
// gcc -o deque_pointers deque_pointers.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int val;
    struct Node *prev, *next;
} Node;

Node *dq_head = NULL, *dq_tail = NULL;
int dq_size = 0;

Node *make_node(int val) {
    Node *n = malloc(sizeof(Node));
    n->val = val; n->prev = n->next = NULL;
    return n;
}

void push_front(int val) {
    Node *node = make_node(val);
    if (!dq_head) { dq_head = dq_tail = node; }
    else { node->next = dq_head; dq_head->prev = node; dq_head = node; }
    dq_size++;
    printf("push_front: %d\n", val);
}

void push_back(int val) {
    Node *node = make_node(val);
    if (!dq_tail) { dq_head = dq_tail = node; }
    else { node->prev = dq_tail; dq_tail->next = node; dq_tail = node; }
    dq_size++;
    printf("push_back: %d\n", val);
}

int pop_front(void) {
    if (!dq_head) { printf("deque empty\n"); return 0; }
    Node *node = dq_head;
    int val = node->val;
    dq_head = node->next;
    if (dq_head) dq_head->prev = NULL;
    else dq_tail = NULL;
    free(node); dq_size--;
    printf("pop_front: %d\n", val);
    return val;
}

int pop_back(void) {
    if (!dq_tail) { printf("deque empty\n"); return 0; }
    Node *node = dq_tail;
    int val = node->val;
    dq_tail = node->prev;
    if (dq_tail) dq_tail->next = NULL;
    else dq_head = NULL;
    free(node); dq_size--;
    printf("pop_back: %d\n", val);
    return val;
}

void print_deque(void) {
    printf("deque:");
    for (Node *c = dq_head; c; c = c->next) printf(" %d", c->val);
    printf("\n");
}

int main(void) {
    push_front(1); push_front(2); push_back(3); push_back(4);
    print_deque();
    printf("size: %d\n", dq_size);
    pop_front(); pop_back();
    print_deque();
    printf("size: %d\n", dq_size);
    while (dq_head) pop_front();
    printf("deque cleared\n");
    return 0;
}

File: source/javascript/deque_pointers.js
// Double-ended queue using JS objects (simulating pointer-based deque)
let dqHead = null, dqTail = null, dqSize = 0;

function makeNode(val) { return { val, prev: null, next: null }; }

function pushFront(val) {
    const node = makeNode(val);
    if (dqHead === null) { dqHead = node; dqTail = node; }
    else { node.next = dqHead; dqHead.prev = node; dqHead = node; }
    dqSize++;
    console.log(`push_front: ${val}`);
}

function pushBack(val) {
    const node = makeNode(val);
    if (dqTail === null) { dqHead = node; dqTail = node; }
    else { node.prev = dqTail; dqTail.next = node; dqTail = node; }
    dqSize++;
    console.log(`push_back: ${val}`);
}

function popFront() {
    if (dqHead === null) { console.log("deque empty"); return null; }
    const node = dqHead;
    const val = node.val;
    const nxt = node.next;
    if (nxt !== null) nxt.prev = null;
    else dqTail = null;
    dqHead = nxt;
    dqSize--;
    console.log(`pop_front: ${val}`);
    return val;
}

function popBack() {
    if (dqTail === null) { console.log("deque empty"); return null; }
    const node = dqTail;
    const val = node.val;
    const prv = node.prev;
    if (prv !== null) prv.next = null;
    else dqHead = null;
    dqTail = prv;
    dqSize--;
    console.log(`pop_back: ${val}`);
    return val;
}

function printDeque() {
    let cur = dqHead;
    const parts = [];
    while (cur !== null) { parts.push(cur.val); cur = cur.next; }
    process.stdout.write("deque:");
    for (const v of parts) process.stdout.write(` ${v}`);
    process.stdout.write("\n");
}

pushFront(1); pushFront(2); pushBack(3); pushBack(4);
printDeque();
console.log(`size: ${dqSize}`);

popFront(); popBack();
printDeque();
console.log(`size: ${dqSize}`);

while (dqHead !== null) popFront();
console.log("deque cleared");
